File: offer56-II-shu-zu-zhong-shu-zi-chu-xian-de-ci-shu-ii-lcof/python/Solution.py
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        n = len(nums)
        bits = [0] * 32
        mask = 1
        for i in range(32):
            for j in range(n):
                if (nums[j]&mask) != 0:
                    bits[i] = (bits[i] + 1) % 3
            mask <<= 1
        result = 0
        mask = 1
        for i in range(32):
            if bits[i] == 1:
                result += mask
            mask <<= 1
        return result

File: offer56-II-shu-zu-zhong-shu-zi-chu-xian-de-ci-shu-ii-lcof/cpp/Solution.cpp
class Solution {
public:
    int singleNumber(vector<int>& nums) {
        int n = nums.size();
        vector<int> bits(32);
        long mask = 1;
        for (int i = 0; i < 32; ++i) {
            for (int j = 0; j < n; ++j) {
                if ((nums[j] & mask) != 0) {
                    bits[i] = (bits[i] + 1) % 3;
                }
            }
            mask <<= 1;
        }
        int result = 0;
        mask = 1;
        for (int i = 0; i < 32; ++i) {
            if (bits[i] == 1) {
                result += mask;
            }
            mask <<= 1;
        }
        return result;;
    }
};
